pkg/template/event_test.go

```diff
@@ -0,0 +1,309 @@
+	}
+	paramEntireEventEventPathVar = pipelinev1.Param{
+		Name:  "paramEntireEventEventPathVar",
+		Value: pipelinev1.ArrayOrString{StringVal: "bar-$(event)-bar"},
+	}
+	wantParamEntireEventEventPathVar = pipelinev1.Param{
+		Name:  "paramEntireEventEventPathVar",
+		Value: pipelinev1.ArrayOrString{StringVal: `bar-{"one": "onevalue", "two": {"two": "twovalue"}, "three": {"three": {"three": {"three": {"three": "threevalue"}}}}}-bar`},
+	}
+	paramOneBogusEventPathVar = pipelinev1.Param{
+		Name:  "paramOneBogusEventPathVar",
+		Value: pipelinev1.ArrayOrString{StringVal: "bar-$(event.bogus.path)-bar"},
+	}
+	wantParamOneBogusEventPathVar = pipelinev1.Param{
+		Name:  "paramOneBogusEventPathVar",
+		Value: pipelinev1.ArrayOrString{StringVal: "bar-$(event.bogus.path)-bar"},
+	}
+	paramMultipleBogusEventPathVars = pipelinev1.Param{
+		Name:  "paramMultipleBogusEventPathVars",
+		Value: pipelinev1.ArrayOrString{StringVal: "bar-$(event.bogus.path)-$(event.two.bogus)-$(event.three.bogus)-bar"},
+	}
+	wantParamMultipleBogusEventPathVars = pipelinev1.Param{
+		Name:  "paramMultipleBogusEventPathVars",
+		Value: pipelinev1.ArrayOrString{StringVal: "bar-$(event.bogus.path)-$(event.two.bogus)-$(event.three.bogus)-bar"},
+	}
+)
+
+func Test_applyEventToParam(t *testing.T) {
+	type args struct {
+		event []byte
+		param pipelinev1.Param
+	}
+	tests := []struct {
+		// name    string
+		args    args
+		want    pipelinev1.Param
+		wantErr bool
+	}{
+		{
+			// name: "no eventpath variable",
+			args:    args{event: []byte{}, param: paramNoEventPathVar},
+			want:    wantParamNoEventPathVar,
+			wantErr: false,
+		},
+		{
+			// name: "one eventpath variable",
+			args:    args{event: testEventJSON, param: paramOneEventPathVar},
+			want:    wantParamOneEventPathVar,
+			wantErr: false,
+		},
+		{
+			// name: "multiple identical eventpath variables",
+			args:    args{event: testEventJSON, param: paramMultipleIdenticalEventPathVars},
+			want:    wantParamMultipleIdenticalEventPathVars,
+			wantErr: false,
+		},
+		{
+			// name: "multiple unique eventpath variables",
+			args:    args{event: testEventJSON, param: paramMultipleUniqueEventPathVars},
+			want:    wantParamMultipleUniqueEventPathVars,
+			wantErr: false,
+		},
+		{
+			// name: "entire event eventpath variable",
+			args:    args{event: testEventJSON, param: paramEntireEventEventPathVar},
+			want:    wantParamEntireEventEventPathVar,
+			wantErr: false,
+		},
+		{
+			// name: "json subobject",
+			args:    args{event: testEventJSON, param: paramSubobjectEventPathVar},
+			want:    wantParamSubobjectEventPathVar,
+			wantErr: false,
+		},
+		{
+			// name: "error one eventpath not found",
+			args:    args{event: testEventJSON, param: paramOneBogusEventPathVar},
+			wantErr: true,
+			want:    wantParamOneBogusEventPathVar,
+		},
+		{
+			// name: "error multiple eventpath not found",
+			args:    args{event: testEventJSON, param: paramMultipleBogusEventPathVars},
+			wantErr: true,
```

Comment:
i am generally not a fan of using a boolean like `wantErr` in a table driven test like this - if a test case results in an error, the logic in the test is usually pretty different from the logic if it succeeds so it seems easier to write, read and maintain if instead we have a completely separate test for errors, so something like `Test_applyEventToParam` and `Test_applyEventToParam_errors`

Reply:
Interesting, ok. Thanks for the advice 👍